synapse/replication/tcp/handler.py

```diff
@@ -0,0 +1,277 @@
+            rows: a list of Stream.ROW_TYPE objects as returned by
+                Stream.parse_row.
+        """
+        logger.debug("Received rdata %s -> %s", stream_name, token)
+        await self._replication_data_handler.on_rdata(stream_name, token, rows)
+
+    async def on_POSITION(self, cmd: PositionCommand):
+        stream = self._streams.get(cmd.stream_name)
+        if not stream:
+            logger.error("Got POSITION for unknown stream: %s", cmd.stream_name)
+            return
+
+        # We're about to go and catch up with the stream, so mark as connecting
+        # to stop RDATA being handled at the same time by removing stream from
+        # list of connected streams. We also clear any batched up RDATA from
+        # before we got the POSITION.
+        self._streams_connected.discard(cmd.stream_name)
+        self._pending_batches.clear()
+
+        # We protect catching up with a linearizer in case the replicaiton
+        # connection reconnects under us.
+        with await self._position_linearizer.queue(cmd.stream_name):
+            # Find where we previously streamed up to.
+            current_token = self._replication_data_handler.get_streams_to_replicate().get(
+                cmd.stream_name
+            )
+            if current_token is None:
+                logger.warning(
+                    "Got POSITION for stream we're not subscribed to: %s",
+                    cmd.stream_name,
+                )
+                return
+
+            # Fetch all updates between then and now.
+            limited = True
+            while limited:
+                updates, current_token, limited = await stream.get_updates_since(
+                    current_token, cmd.token
+                )
+                if updates:
+                    await self.on_rdata(
+                        cmd.stream_name,
+                        current_token,
+                        [stream.parse_row(update[1]) for update in updates],
+                    )
+
+            # We've now caught up to position sent to us, notify handler.
+            await self._replication_data_handler.on_position(cmd.stream_name, cmd.token)
+
+        self._streams_connected.add(cmd.stream_name)
+
+        # Handle any RDATA that came in while we were catching up.
+        rows = self._pending_batches.pop(cmd.stream_name, [])
+        if rows:
+            # We need to make sure we filter out RDATA rows with a token less
+            # than what we've caught up to. This is slightly fiddly because of
+            # "batched" rows which have a `None` token, indicating that they
+            # have the same token as the next row with a non-None token.
+            #
+            # We do this by walking the list backwards, first removing any RDATA
+            # rows that are part of an uncompeted batch, then taking rows while
+            # their token is either None or greater than where we've caught up
+            # to.
```

Comment:
Actually, am I overthinking this? Should we just pass on all RDATA commands that we get *after* the POSITION? I.e., clearing `pending_batches`  when we receive a POSITION?

Reply:
It looks like that's already happening (https://github.com/matrix-org/synapse/pull/7185/files/5104d1673bb4a3be3bd2a655dfde568fda01226c..534bd868e50cd1fe2efd52d4ec0ec92452ac6a6b#diff-bff709ecab561a0aa9f155333fcc1b0dR127), but I think there's a slight problem here:

* we receive a POSITION and start a catchup
* RDATA arrives
* connection drops and we reconnect
* we receive another POSITION, clear `pending_batches`, and start waiting for the linearizer
* more RDATA arrives and gets added to `pending_batches`
* the first catchup completes, and we process all the RDATAs which arrived since the second POSITION despite having not caught up to it.

At this point, this stuff feels very much a separate problem to "Move client command handling out of TCP protocol". I've also realised that the current impl is also racy af, which can't be helping with #7206.

Reply:
Agreed, I'll revert to previous handling and we can think about this separately.